docs/proposals/PROPOSAL_H.md

```diff
@@ -108,103 +110,172 @@ Schema:
   "compatibilities": [
     {
-      "cpu-model.vendor_id": "GenuineIntel",
-      "cpu-cpuid.AMXFP16": "true",
-      "kernel-selinux.enabled": "false",
-      "kernel-config.PREEMPT": "true",
-      "kernel-version.full": "[4.19, 5.16)",
-      "system.glibc": "[2.31, 2.37]",
-      "system.openmpi": "[4.1.6, 4.1.6]",
-      "pci-15B3-019D.present": "true",
+      "oci.cpu.vendor": "GenuineIntel",
+      "oci.cpu.features": "AMXFP16,",
+      "oci.kernel.selinux": "false"
+      "oci.kernel.configurations": "PREEMPT,",
+      "oci.kernel.versions": "[4.19, 5.16)",
+      "oci.os.glibc": "[2.31, 2.37]",
+      "oci.os.openmpi": "[4.1.6, 4.1.6]",
+      "oci.pci.devices": "15B3.020D,",
       "compatibility-score": 50,
     },
     {
-      "cpu-model.vendor_id": "GenuineIntel",
-      "cpu-cpuid.AVX512FP16": "true",
-      "kernel-selinux.enabled": "false",
-      "kernel-config.PREEMPT": "true",
-      "kernel-version.full": "[4.19, 5.16)",
-      "system.glibc": "[2.31, 2.37]",
-      "system.openmpi": "[4.1.6, 4.1.6]",
-      "pci-15B3-020D.present": "true",
-      "compatibility-score": 100,
+      "oci.cpu.vendor": "GenuineIntel",
+      "oci.cpu.features": "AVX512FP16,",
+      "oci.kernel.selinux": "false",
+      "oci.kernel.configurations": "PREEMPT,",
+      "oci.kernel.versions": "[4.19, 5.16)",
+      "oci.os.glibc": "[2.31, 2.37]",
+      "oci.os.openmpi": "[4.1.6, 4.1.6]",
+      "oci.pci.devices": "15B3.020D,",
+      "oci.compatibility-score": 100,
     },
     {
-      "cpu-model.vendor_id": "AuthenticAMD",
-      "cpu-cpuid.FPHP": "true",
-      "kernel-selinux.enabled": "false",
-      "kernel-config.PREEMPT": "true",
-      "kernel-version.full": "[4.19, 5.16)",
-      "system.glibc": "[2.31, 2.37]",
-      "system.openmpi": "[4.1.6, 4.1.6]",
-      "pci-1002-67ff.present": "true",
-      "compatibility-score": 80,
+      "oci.cpu.vendor": "AuthenticAMD",
+      "oci.cpu.features": "FPHP,",
+      "oci.kernel.selinux": "false",
+      "oci.kernel.configurations": "PREEMPT,",
+      "oci.kernel.versions": "[4.19, 5.16)",
+      "oci.os.glibc": "[2.31, 2.37]",
+      "oci.os.openmpi": "[4.1.6, 4.1.6]",
+      "oci.pci.devices": "1002.67ff,",
+      "oci.compatibility-score": 80,
       "annotation": "only work with AMD CPU and AMD GPU"
     }
   ]
 }
 ```
 
-## Alternative place to hold compatibility set
+## Compatibilities Embeded Inside Platform Object
 
-Considering the compatibilities are composed of a list of compatibility set, just similar as the idea in proposal E, the platform.features could be reused.
+### Extend Platform Object with New Compatibilities Field
 
-Each compatibility set will be encoded into a string, then the platform.features can hold all the compatibility artifact content, the downside is the platform.features content is hard to read for human.
+Add a new field in platform object in [image index](https://github.com/opencontainers/image-spec/blob/main/image-index.md#image-index-property-descriptions).
```

Comment:
Is there an upper size limit envisioned for the compatibilities list of maps? I ask because there is a size limit recommended for an Index (4 MiB) which would include a list of images, each with their own platform field.

Reply:
Ok, will add size limit here.

Reply:
I think I may want to look at some real world examples to get a better feel for what this will do to images. Often people see what they want from a proposal, and don't realize how others intend to implement it. If a typical Index goes from 1KB to >500KB to store multiple image references with lots of extra compatibilities data, I could see registry operators getting concerned.

For some of these registries, manifests are maintained in a separate database from the blobs. Their manifest DB is designed for frequent access of small but often queried and changing data (latest tag constantly being updated).

Reply:
Introducing compatibilities field in platform is very controversial. I doubt I should include it in the proposal. The original motivation is to save API requests to registriy.

For registry operator, they may prefer to scale out the registry to provide additonal artifact API request and json blob download. Storing big compatibilities in manifest DB will change the traffic pattern, I don't think SRE/maintainance engineer like it much.

Reply:
In general, I'm in favor of having a way for runtimes to quickly select the image to run. One index manifest pull is much less overhead than querying for a compatibility artifact for every entry in the index. I just want to be careful that we don't get bloat from data that is unrelated to that purpose.

Reply:
Some thoughts about the size impact of platform.compatibilities:
1) set size limit for platform.compatibilities
2) use short label name
3) use even shorter label name, and a dict to map the shorter name to full name, tools can use the dict to output human readable information.

Reply:
One simple check is if every entry in the index has the same compatibilities entry (for something like the kernel, libc, etc), then the value isn't needed. You would only want to include compatibility entries that differentiate one entry from the others.

Reply:
> One simple check is if every entry in the index has the same compatibilities entry (for something like the kernel, libc, etc), then the value isn't needed. You would only want to include compatibility entries that differentiate one entry from the others.

it's a little difficult to find a good way to implement it (some entry without value like "oci.os.glibc": "" )

Reply:
I don't think we'd want any redundancy, between any two places. If there is an OCI maintained label, it should be present in one (and only one) place.

Reply:
Also I think we should move labels discussion into a next step of proposals. We are mostly concerned about structure here.

Reply:
> > One simple check is if every entry in the index has the same compatibilities entry (for something like the kernel, libc, etc), then the value isn't needed. You would only want to include compatibility entries that differentiate one entry from the others.
> 
> it's a little difficult to find a good way to implement it (some entry without value like "oci.os.glibc": "" )

We may be thinking of different problems. When writing the index, every entry in the index is available to the writer. It would be a dedup step, performed across all the entries of the index. If one compatibility key+value pair exists in every entry of the index for a given os/arch/version, remove it from the index, because there's no added value.

Reply:
> We may be thinking of different problems. When writing the index, every entry in the index is available to the writer. It would be a dedup step, performed across all the entries of the index. If one compatibility key+value pair exists in every entry of the index for a given os/arch/version, remove it from the index, because there's no added value.

That makes sense to me. The tools that create the index would have to perform the dedup step, right?

Reply:
> That makes sense to me. The tools that create the index would have to perform the dedup step, right?

"Have" is a strong word, the spec language would probably end up being a "SHOULD", if not a looser "implementers guidance". But otherwise yes, the tool creating the index has everything needed to implement this without generating too much noise.

Reply:
OK, deduplication will be updated in next commit